Add symbolic shape comparison optimization

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64323 Add embedding shape analysis
* #64121 Max Pool with indices
* #63530 Add Maxpool to shape analysis / Opinfo
* **#64300 Add symbolic shape comparison optimization**
* #64299 Refactor to use shape arguments
* #63516 Add view with negative dim
* #63615 Generalize expand logic
* #63407 Add permute, arange
* #63365 Add support for slice, selec twith int, index_select
* #63099 Add squeeze, unsqueeze, transpose shape functins
* #63050 Add batch of unary functions

Differential Revision: [D30738146](https://our.internmc.facebook.com/intern/diff/D30738146)

Commit:
Add symbolic shape comparison optimization

[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -369,6 +367,36 @@ struct SymbolicShapeAnalyzer {
             }
+            for (const auto& sym_uses: use.user->output()->uses()) {
+              auto k = sym_uses.user->kind();
+              if (k != aten::ge && k != aten::le && k != aten::ne && k != aten::eq) {
```
why can't we do `gt` and `lt`?

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

Commit:
Update on "Add symbolic shape comparison optimization"




[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738146).

Commit:
Update on "Add symbolic shape comparison optimization"


Differential Revision: [D30738146](https://our.internmc.facebook.com/intern/diff/D30738146)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738146).

Commit:
Update on "Add symbolic shape comparison optimization"


Differential Revision: [D30738146](https://our.internmc.facebook.com/intern/diff/D30738146)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738146).
